File: src/test/java/com/baulsupp/oksocial/services/test/TestAuthInterceptor.java
package com.baulsupp.oksocial.services.test;

import com.baulsupp.oksocial.apidocs.ApiDocPresenter;
import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.authenticator.ValidatedCredentials;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token;
import com.baulsupp.oksocial.credentials.ServiceDefinition;
import com.baulsupp.oksocial.output.OutputHandler;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.Future;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import static java.util.concurrent.CompletableFuture.completedFuture;

public class TestAuthInterceptor implements AuthInterceptor<Oauth2Token> {
  @Override public Response intercept(Interceptor.Chain chain, Oauth2Token credentials)
      throws IOException {
    return chain.proceed(chain.request());
  }

  @Override public Oauth2Token authorize(OkHttpClient client, OutputHandler outputHandler,
      List<String> authArguments)
      throws IOException {
    if (authArguments.isEmpty()) {
      return new Oauth2Token("testToken");
    } else {
      return new Oauth2Token(authArguments.get(0));
    }
  }

  @Override public Future<Optional<ValidatedCredentials>> validate(OkHttpClient client,
      Request.Builder requestBuilder, Oauth2Token credentials) throws IOException {
    return completedFuture(Optional.empty());
  }

  @Override public ServiceDefinition<Oauth2Token> serviceDefinition() {
    return new Oauth2ServiceDefinition("localhost", "Test Service", "test",
        "https://docs.test.com");
  }

  @Override public Collection<String> hosts() {
    return Arrays.asList("test.com", "api1.test.com");
  }

  @Override public ApiDocPresenter apiDocPresenter(String url) {
    return (url1, outputHandler, client) -> outputHandler.info("Test: " + url1);
  }
}

File: src/test/kotlin/com/baulsupp/oksocial/services/test/TestAuthInterceptor.kt
package com.baulsupp.oksocial.services.test

import com.baulsupp.oksocial.Token
import com.baulsupp.oksocial.apidocs.ApiDocPresenter
import com.baulsupp.oksocial.authenticator.AuthInterceptor
import com.baulsupp.oksocial.authenticator.ValidatedCredentials
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token
import com.baulsupp.oksocial.credentials.ServiceDefinition
import com.baulsupp.oksocial.output.OutputHandler
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response

class TestAuthInterceptor : AuthInterceptor<Oauth2Token>() {

  override fun intercept(chain: Interceptor.Chain, credentials: Oauth2Token): Response = chain.proceed(chain.request())

  override suspend fun authorize(client: OkHttpClient, outputHandler: OutputHandler<Response>, authArguments: List<String>): Oauth2Token =
      if (authArguments.isEmpty()) {
        Oauth2Token("testToken")
      } else {
        Oauth2Token(authArguments[0])
      }

  override suspend fun validate(client: OkHttpClient, credentials: Oauth2Token): ValidatedCredentials =
      ValidatedCredentials("aaa", null)

  override fun serviceDefinition(): ServiceDefinition<Oauth2Token> =
      Oauth2ServiceDefinition("localhost", "Test Service", "test", "https://docs.test.com", "https://apps.test.com")

  override fun hosts(): Set<String> = setOf("test.com", "api1.test.com")

  override fun apiDocPresenter(url: String): ApiDocPresenter = object : ApiDocPresenter {
    override suspend fun explainApi(url: String, outputHandler: OutputHandler<Response>, client: OkHttpClient, tokenSet: Token) {
      outputHandler.info("Test: $url")
    }
  }
}
